x/ecocredit/server/core/msg_server.go

```diff
@@ -0,0 +1,1200 @@
+	//	BuyOrderId:         orderID,
+	//	Selection:          &selection,
+	//	Quantity:           o.Quantity,
+	//	BidPrice:           o.BidPrice,
+	//	DisableAutoRetire:  o.DisableAutoRetire,
+	//	DisablePartialFill: o.DisablePartialFill,
+	//	Expiration:         o.Expiration,
+	//})
+	//return orderID, err
+}
+
+// AllowAskDenom adds a new ask denom
+// TODO: impl with ORM https://github.com/regen-network/regen-ledger/issues/728
+func (s serverImpl) AllowAskDenom(ctx context.Context, denom *marketplacev1beta1.MsgAllowAskDenom) (*marketplacev1beta1.MsgAllowAskDenomResponse, error) {
+	panic("implement me")
+	// ctx := types.UnwrapSDKContext(goCtx)
+	//
+	//rootAddress := s.accountKeeper.GetModuleAddress(govtypes.ModuleName).String()
+	//
+	//if req.RootAddress != rootAddress {
+	//	return nil, sdkerrors.ErrUnauthorized.Wrapf("root address must be governance module address, got: %s, expected: %s", req.RootAddress, rootAddress)
+	//}
+	//
+	//err := s.askDenomTable.Create(ctx, &ecocredit.AskDenom{
+	//	Denom:        req.Denom,
+	//	DisplayDenom: req.DisplayDenom,
+	//	Exponent:     req.Exponent,
+	//})
+	//if err != nil {
+	//	return nil, err
+	//}
+	//
+	//err = ctx.EventManager().EmitTypedEvent(&ecocredit.EventAllowAskDenom{
+	//	Denom:        req.Denom,
+	//	DisplayDenom: req.DisplayDenom,
+	//	Exponent:     req.Exponent,
+	//})
+	//if err != nil {
+	//	return nil, err
+	//}
+	//
+	//return &ecocredit.MsgAllowAskDenomResponse{}, nil
+}
+
+func (s serverImpl) CreateBasket(ctx context.Context, basket *basketv1beta1.MsgCreateBasket) (*basketv1beta1.MsgCreateBasketResponse, error) {
+	panic("implement me")
+}
+
+func (s serverImpl) AddToBasket(ctx context.Context, basket *basketv1beta1.MsgAddToBasket) (*basketv1beta1.MsgAddToBasketResponse, error) {
+	panic("implement me")
+}
+
+func (s serverImpl) TakeFromBasket(ctx context.Context, basket *basketv1beta1.MsgTakeFromBasket) (*basketv1beta1.MsgTakeFromBasketResponse, error) {
+	panic("implement me")
+}
+
+func (s serverImpl) PickFromBasket(ctx context.Context, basket *basketv1beta1.MsgPickFromBasket) (*basketv1beta1.MsgPickFromBasketResponse, error) {
+	panic("implement me")
+}
+
+// ------- UTILITIES ------
+
+// Checks if the given address is in the allowlist of credit class creators
+func (s serverImpl) isCreatorAllowListed(ctx sdk.Context, allowlist []string, designer sdk.Address) bool {
+	for _, addr := range allowlist {
+		ctx.GasMeter().ConsumeGas(gasCostPerIteration, "credit class creators allowlist")
+		allowListedAddr, _ := sdk.AccAddressFromBech32(addr)
+		if designer.Equals(allowListedAddr) {
+			return true
+		}
+	}
+	return false
+}
+
+// assertClassIssuer makes sure that the issuer is part of issuers of given classID.
+// Returns ErrUnauthorized otherwise.
+func (s serverImpl) assertClassIssuer(goCtx context.Context, classID, issuer string) error {
+	it, err := s.stateStore.ClassIssuerStore().List(goCtx, ecocreditv1beta1.ClassIssuerClassIdIssuerIndexKey{}.WithClassId(classID))
```

Comment:
we don't need to iterate, we can just do a `Has(classId, issuer)` I'm pretty sure

Reply:
ah right, fixed it.